### DISCUSSION TITLE
Testing an NgRx project

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#357] [Announcements]
# Testing an NgRx project

How write maintainable NgRx tests within an Angular application

https://timdeschryver.dev/blog/testing-an-ngrx-project

**Comment by rafal21-ops** [comment]
Great article. I definitely want to try that in practise and do compare between TestBed option and approach you present. I have some question about ngrx.
1. Do you split up selector file into smaller files. For example selector file which only retrieve raw data. Selector file that contain logic. Do you keep that files inside selectors directory then or have different approach?
2. Do you group effect into separate files based on effects which operate on router, dialogs, do only http calls and these which handle flow of others effect? Or maybe this only complicate things?
Thanks.

**Reply by timdeschryver** [reply]
Thanks @rafal21-ops , to answer your questions:

1. It depends. Usually the selector contains the logics. If the logic is complex it might happen that I create a helper function/file with the calculations. This can also be useful to test individual "pieces" in isolation. Depending on the project/team we might also end up with selectors files next to components.
2. I usually add them in one class (functional/domain grouping over technical grouping). If you want to split effects into mulitple files I suggest you to take a look at [functional effects](https://ngrx.io/guide/effects#functional-effects)
